### PR TITLE
[Plots.Segment] Implementation

### LINKED ISSUE #2394: Line Segment Plot
7fa1ff70-6389-47a2-9d55-023acd396d9f

### PR DESCRIPTION
**Pull request by rcchen**
image

http://jsfiddle.net/qLjusc6w/

Fixes #2394

### CODE REVIEW

**Comment by jtlan** [comment]
Our modern art repertoire expands!

File: src/drawers/segmentDrawer.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+export module Drawers {
+  export class Segment extends Drawer {
+
+    constructor(dataset: Dataset) {
+      super(dataset);
+      this._svgElementName = "line";
+    }
+
+    protected _applyDefaultAttributes(selection: d3.Selection<any>) {
+      super._applyDefaultAttributes(selection);
```
Comment (jtlan) [inline src/drawers/segmentDrawer.ts]: What's the point of this override?

File: src/plots/segmentPlot.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+      var attrToProjector = super._generateAttrToProjector();
+      return attrToProjector;
+    }
+
+    protected _generateDrawSteps(): Drawers.DrawStep[] {
+      return [{attrToProjector: this._generateAttrToProjector(), animator: this._getAnimator(Plots.Animator.MAIN)}];
+    }
+
+    public x2(): AccessorScaleBinding<X, number>;
+    public x2(x2: number | Accessor<number> | X | Accessor<X>): Plots.Segment<X, Y>;
```
Comment (jtlan) [inline src/plots/segmentPlot.ts]: Docs.

File: src/plots/segmentPlot.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+      this.addClass("segment-plot");
+    }
+
+    protected _createDrawer(dataset: Dataset) {
+      return new Drawers.Segment(dataset);
+    }
+
+    protected _generateAttrToProjector() {
+      var attrToProjector = super._generateAttrToProjector();
+      return attrToProjector;
```
Comment (jtlan) [inline src/plots/segmentPlot.ts]: What's the point of this override?

**Comment by jtlan** [comment]
Example Fiddle?

File: src/plots/segmentPlot.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    protected _generateDrawSteps(): Drawers.DrawStep[] {
+      return [{attrToProjector: this._generateAttrToProjector(), animator: this._getAnimator(Plots.Animator.MAIN)}];
```
Comment (jtlan) [inline src/plots/segmentPlot.ts]: Not sure this looks quite right. We should stick with an `Animators.Null` until we get a spec from design for what this should look like.

File: src/plots/segmentPlot.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+      this.render();
+      return this;
+    }
+
+    protected _propertyProjectors(): AttributeToProjector {
+      var attrToProjector = super._propertyProjectors();
+      attrToProjector["x1"] = Plot._scaledAccessor(this.x());
+      attrToProjector["x2"] = Plot._scaledAccessor(this.x2());
+      attrToProjector["y1"] = Plot._scaledAccessor(this.y());
+      attrToProjector["y2"] = Plot._scaledAccessor(this.y2());
```
Comment (jtlan) [inline src/plots/segmentPlot.ts]: This implies that `x2` and `y2` are mandatory, unlike on `Plots.Rectangle`. Is that true?

**Comment by rcchen** [comment]
In which I once again forget to remove stuff that was added when hacking out the initial impl...

File: src/plots/segmentPlot.ts (hunk 5 above)
Comment (rcchen) [inline src/plots/segmentPlot.ts]: I implemented it as mandatory, but now come to think of it, it is possible for either one to be optional (not both though, as you end up with a point).

File: src/plots/segmentPlot.ts [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+     * @param {number|Accessor<number>|Y|Accessor<Y>} y2
+     * @returns {Plots.Segment} The calling Segment Plot
+     */
+    public x2(x2: number | Accessor<number> | X | Accessor<X>): Plots.Segment<X, Y>;
+    public x2(x2?: number | Accessor<number> | X | Accessor<X>): any {
+      if (x2 == null) {
+        return this._propertyBindings.get(Segment._X2_KEY);
+      }
+      var xBinding = this.x();
+      var xScale = xBinding && xBinding.scale;
```
Comment (jtlan) [inline src/plots/segmentPlot.ts]: If the user sets `x2()` before `x()`, they won't get the `Scale`, although I'll note that this is an open bug on `Plots.Rectangle` as well.

File: src/plots/segmentPlot.ts [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+     * @param {number|Accessor<number>|Y|Accessor<Y>} y2
+     * @returns {Plots.Segment} The calling Segment Plot.
+     */
+    public y2(y2: number | Accessor<number> | Y | Accessor<Y>): Plots.Segment<X, Y>;
+    public y2(y2?: number | Accessor<number> | Y | Accessor<Y>): any {
+      if (y2 == null) {
+        return this._propertyBindings.get(Segment._Y2_KEY);
+      }
+      var yBinding = this.y();
+      var yScale = yBinding && yBinding.scale;
```
Comment (jtlan) [inline src/plots/segmentPlot.ts]: Same `Scale` problem here.

File: src/plots/segmentPlot.ts [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+     * @returns {Plots.Segment} The calling Segment Plot
+     */
+    public x2(x2: number | Accessor<number> | X | Accessor<X>): Plots.Segment<X, Y>;
+    public x2(x2?: number | Accessor<number> | X | Accessor<X>): any {
+      if (x2 == null) {
+        return this._propertyBindings.get(Segment._X2_KEY);
+      }
+      var xBinding = this.x();
+      var xScale = xBinding && xBinding.scale;
+
```
Comment (jtlan) [inline src/plots/segmentPlot.ts]: Whitespace here, but not in `y2()`?

File: plottable.css [hunk 9]
```diff
@@ -158,3 +158,8 @@ svg.plottable {
   stroke: #CCC;
   stroke-width: 1px;
 }
+
+.plottable .segment-plot line {
+  stroke: #555;
+  stroke-width: 1px;
```
Comment (jtlan) [inline plottable.css]: Do we have design sign-off on this? I would instead default to the same color other `Plot`s do, using an `attr()` call in the `constructor`.

**Comment by crmorford** [comment]
Is there anything we can do about the different width of different segments? Seems like an aliasing thing. Compare the first line (furthest left) to any other line.
screen shot 2015-06-25 at 3 43 00 pm

**Comment by rcchen** [comment]
Dunno if it's avoidable. http://bl.ocks.org/mbostock/1322948

**Comment by jtlan** [comment]
It'd be `grunt eslint` now right?

**Comment by crmorford** [comment]
+1 pending on me...awkward

**Pull request merged**